# THE BRIDGE (Data Science Bootcamp - Agosto 2020)
##  PROYECTO INDIVIDUAL VERANO 2020 
________________
### Análisis de datos asociados a la oferta de empleo en Data Science

Fuente de datos (Data Set)  https://www.kaggle.com/andresionek/data-jobs-listings-glassdoor

Elaborado por : 

- ** Mónica Villasuso  **    villasuso.monica@gmail.com  git @MVillasuso
_________________

# HIPÓTESIS
"Al menos la mitad (50%) de la oferta laboral en el área de Ciencia de Datos corresponde a los perfiles de Análisis (Datos y Negocio)"

## 1. Librerías requeridas para el análisis (* Import *)
_________

In [1]:
import sys 
sys.path.append('../src/utils') 
import pandas as pd 
import numpy as np
from datetime import datetime
import folders_tb as ftb
import visualization_tb as vtb
import mining_data_tb as mtb

import plotly.express as px 

## 2. Carga de datos inicial, de la fuente, en el dataframe        

____________

### Información global de Glassdoor.com   
** Actualizado a Dic 2019 **  
Dataset con búsquedas para los siguientes perfiles

* data-scientist		**
* software-engineer
* data-analyst			**
* research-scientist
* business-analyst		**
* product-manager
* project-manager
* data-engineer		**
* statistician
* dba
* database-engineer
* machine-learning-engineer

https://www.kaggle.com/andresionek/data-jobs-listings-glassdoor

In [2]:
GlassD_df= pd.read_csv("glassdoor.csv", sep=",")


In [3]:
cnames_df = pd.read_csv("country_names_2_digit_codes.csv", sep =",")

_____________
### Información de Glassdoor en USA.   
** Actualizado a Julio 2020 **  
Cada dataset incluye las búsquedas para un determinado perfil, a saber: Data Analyst, Business Analyst, Data Engineer y Data Scientist
- https://www.kaggle.com/andrewmvd/data-scientist-jobs
- https://www.kaggle.com/andrewmvd/data-analyst-jobs
- https://www.kaggle.com/andrewmvd/data-engineer-jobs
- https://www.kaggle.com/andrewmvd/business-analyst-jobs

In [4]:

DA_df = pd.read_csv("DataAnalyst.csv", sep=",")
BA_df = pd.read_csv("BusinessAnalyst.csv", sep=",")
DE_df = pd.read_csv("DataEngineer.csv", sep=",")
DS_df = pd.read_csv("DataScientist.csv", sep=",")


## Otras referencias consultadas
https://www.simplilearn.com/business-analyst-vs-data-analyst-article
https://blog.udacity.com/2014/11/data-science-job-skills.html

## 3. Preparación de los datos - Data wrangling and Data cleaning
________________
Exploración de los datos del dataframe

### df_GlassD
Dataframe original con la información obtenida de la carga del Data Set

In [5]:
GlassD_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 165290 entries, 0 to 165289
Columns: 163 entries, benefits.benefitRatingDecimal to wwfu
dtypes: bool(11), float64(35), int64(23), object(94)
memory usage: 193.4+ MB


### df_jobs
Creación de un subconjunto del Dataset original para lo cual es necesario:   
- Eliminar las columnas con información no relevante para el análisis  
- Transformar los tipos de datos útiles para el análisis  
- Estandarizar los códigos de países para poder utilizarlos posteriormente


In [6]:

GDjobs_df = mtb.elim_cols(GlassD_df)
GDjobs_df = mtb.transf_cols(GDjobs_df)


*** NOTA: De las 163 columnas originales quedaron  19. ***

In [7]:
GDjobs_df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 165290 entries, 0 to 165289
Data columns (total 19 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   empSize      147081 non-null  object        
 1   easyApply    165290 non-null  bool          
 2   empName      162045 non-null  object        
 3   jobTitle     165290 non-null  object        
 4   jobDate      165290 non-null  datetime64[ns]
 5   salHigh      3662 non-null    float64       
 6   salLow       3662 non-null    float64       
 7   jobDesc      165286 non-null  object        
 8   jobSource    165290 non-null  object        
 9   country      115262 non-null  object        
 10  lat          165290 non-null  float64       
 11  long         165290 non-null  float64       
 12  location     165285 non-null  object        
 13  foundedYear  165290 non-null  int64         
 14  industry     117885 non-null  object        
 15  revenue      147081 non-null  obje

In [8]:
GDjobs_df.describe(include="all")

empSize easyApply empName         jobTitle              jobDate  \
count     147081    165290  162045           165290               165290   
unique         8         2   31114            64844                  248   
top     10000--1     False  Amazon  Project Manager  2019-11-01 00:00:00   
freq       48801    147652    2494             4648                14099   
first        NaN       NaN     NaN              NaN  2017-01-21 00:00:00   
last         NaN       NaN     NaN              NaN  2019-11-10 00:00:00   
mean         NaN       NaN     NaN              NaN                  NaN   
std          NaN       NaN     NaN              NaN                  NaN   
min          NaN       NaN     NaN              NaN                  NaN   
25%          NaN       NaN     NaN              NaN                  NaN   
50%          NaN       NaN     NaN              NaN                  NaN   
75%          NaN       NaN     NaN              NaN                  NaN   
max          NaN       NaN     NaN              NaN                  NaN   

              salHigh         salLow  \
count     3662.000000    3662.000000   
unique            NaN            NaN   
top               NaN            NaN   
freq              NaN            NaN   
first             NaN            NaN   
last              NaN            NaN   
mean    115832.403605   77825.276898   
std      42077.284501   30736.395180   
min         13.000000      10.000000   
25%      85217.500000   55283.500000   
50%     109630.500000   72435.000000   
75%     140000.000000   95966.000000   
max     376948.000000  236650.000000   

                                                  jobDesc  \
count                                              165286   
unique                                             107692   
top     Credit Suisse Careers Credit Suisse Menu Job S...   
freq                                                   33   
first                                                 NaN   
last                                                  NaN   
mean                                                  NaN   
std                                                   NaN   
min                                                   NaN   
25%                                                   NaN   
50%                                                   NaN   
75%                                                   NaN   
max                                                   NaN   

                       jobSource country            lat           long  \
count                     165290  115262  165290.000000  165290.000000   
unique                       217     503            NaN            NaN   
top     Indeed XML Feed Employer      BE            NaN            NaN   
freq                       53202    4587            NaN            NaN   
first                        NaN     NaN            NaN            NaN   
last                         NaN     NaN            NaN            NaN   
mean                         NaN     NaN      26.499080      29.642232   
std                          NaN     NaN      26.013832      62.586654   
min                          NaN     NaN     -77.850000    -171.983300   
25%                          NaN     NaN       2.922800       0.000000   
50%                          NaN     NaN      33.823400      12.133300   
75%                          NaN     NaN      48.866700      77.587300   
max                          NaN     NaN      67.183300     177.366700   

         location    foundedYear                      industry  \
count      165285  165290.000000                        117885   
unique       7074            NaN                           137   
top     Singapore            NaN  Computer Hardware & Software   
freq         5100            NaN                         13971   
first         NaN            NaN                           NaN   
last          NaN            NaN                           NaN   
mean          NaN    1273.873

Estandarización de la información correspondiente al país

In [9]:
GDjobs_df = mtb.norm_country(GDjobs_df, cnames_df)
GDjobs_df = mtb.ubicar_loc(GDjobs_df)

** NOTA: Se agregaron las  columnas ccode y cname con la información estandarizada del país  (código y nombre) **

Asignar  experiencia requerida , nivel del puesto y tipo de trabajo según la información contenida en el campo JobTitle

In [10]:
GDjobs_df = mtb.experience (GDjobs_df)
GDjobs_df = mtb.level (GDjobs_df)
GDjobs_df = mtb.jobType(GDjobs_df)

Completar con el valor NA (Not available) los campos en los que la información no exista (valores Nan,None, nulls)

In [11]:
GDjobs_df = mtb.llenar_na(GDjobs_df)

** RESULTADO DATA WRANGLING:   
- De las 165290 filas del  DF original hemos eliminado (por inconsistencias) alrededor del 5%, quedando 155890 registros 
- Se agregaron 5 columnas con información estandarizada y que facilita la interpretación
- Se redujo la memoria utilizada por el dataset  en más de 80% (De 193 MB a 35 MB)

In [12]:
GDjobs_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 155890 entries, 0 to 165284
Data columns (total 24 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   empSize      155890 non-null  object        
 1   easyApply    155890 non-null  bool          
 2   empName      155890 non-null  object        
 3   jobTitle     155890 non-null  object        
 4   jobDate      155890 non-null  datetime64[ns]
 5   salHigh      3645 non-null    float64       
 6   salLow       3645 non-null    float64       
 7   jobDesc      155890 non-null  object        
 8   jobSource    155890 non-null  object        
 9   country      155890 non-null  object        
 10  lat          155890 non-null  float64       
 11  long         155890 non-null  float64       
 12  location     155890 non-null  object        
 13  foundedYear  155890 non-null  int64         
 14  industry     155890 non-null  object        
 15  revenue      155890 non-null  obje

### res_df
** Creación de  un DF resumen (agrupado por los principales campos) para comenzar el análisis de tendencias los datos **

In [13]:
res_df = mtb.resumen_df(GDjobs_df)

In [14]:
res_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 45860 entries, 0 to 45859
Data columns (total 16 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   size          45860 non-null  object 
 1   sector        45860 non-null  object 
 2   industry      45860 non-null  object 
 3   type          45860 non-null  object 
 4   ccode         45860 non-null  object 
 5   cname         45860 non-null  object 
 6   exp           45860 non-null  object 
 7   level         45860 non-null  object 
 8   jobType       45860 non-null  object 
 9   total         45860 non-null  int64  
 10  salLow_min    1518 non-null   float64
 11  salLow_max    1518 non-null   float64
 12  salLow_mean   1518 non-null   float64
 13  salHigh_min   1518 non-null   float64
 14  salHigh_max   1518 non-null   float64
 15  salHigh_mean  1518 non-null   float64
dtypes: float64(6), int64(1), object(9)
memory usage: 5.6+ MB


In [15]:
res_df['total'].sum()

155890

In [16]:
res_df.exp.value_counts()

N/A           33757
Senior         9768
Junior         1826
Internship      509
Name: exp, dtype: int64

In [17]:
res_df.describe(include="all")

size                  sector industry               type  ccode  \
count      45860                   45860    45860              45860  45860   
unique         9                      26      138                 15    176   
top     10000--1  Information Technology      N/A  Company - Private     DE   
freq       12798                   11308     7700              21920   2320   
mean         NaN                     NaN      NaN                NaN    NaN   
std          NaN                     NaN      NaN                NaN    NaN   
min          NaN                     NaN      NaN                NaN    NaN   
25%          NaN                     NaN      NaN                NaN    NaN   
50%          NaN                     NaN      NaN                NaN    NaN   
75%          NaN                     NaN      NaN                NaN    NaN   
max          NaN                     NaN      NaN                NaN    NaN   

          cname    exp    level jobType         total     salLow_min  \
count     45860  45860    45860   45860  45860.000000    1518.000000   
unique      176      4       13       7           NaN            NaN   
top     Germany    N/A  Manager                   NaN            NaN   
freq       2320  33757    12794   29225           NaN            NaN   
mean        NaN    NaN      NaN     NaN      3.399259   72588.517787   
std         NaN    NaN      NaN     NaN      6.999297   30010.940397   
min         NaN    NaN      NaN     NaN      1.000000      10.000000   
25%         NaN    NaN      NaN     NaN      1.000000   51749.250000   
50%         NaN    NaN      NaN     NaN      2.000000   66335.000000   
75%         NaN    NaN      NaN     NaN      3.000000   89177.250000   
max         NaN    NaN      NaN     NaN    231.000000  236650.000000   

          salLow_max    salLow_mean    salHigh_min    salHigh_max  \
count     1518.00000    1518.000000    1518.000000    1518.000000   
unique           NaN            NaN            NaN            NaN   
top              NaN            NaN            NaN            NaN   
freq             NaN            NaN            NaN            NaN   
mean     79503.86166   75969.732376  108126.714097  117911.237154   
std      32665.31079   29953.417051   40741.883160   44746.302312   
min         10.00000      10.000000      13.000000      14.000000   
25%      55350.50000   53987.750000   79318.500000   85377.250000   
50%      73248.00000   70202.800000  100636.500000  110579.500000   
75%      98009.75000   93414.500000  130369.250000  142501.250000   
max     236650.00000  236650.000000  376376.000000  376948.000000   

         salHigh_mean  
count     1518.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean    112915.739700  
std      40617.312994  
min         14.000000  
25%      83206.406250  
50%     107128.333333  
75%     136312.000000  
max     376376.000000

In [18]:
cjob_df = pd.DataFrame({'jobClas': ['Others', 'Project Mgmt.', 'Data Science','Data Science','Data Science','Data Science','Data Science'] , 
                    "jobType": ['', 'Project Manager', 'Business Analyst', 'Data Scientist','Data Analyst','Data Engineer', 'ML/AI']})

In [19]:
cjob_df

jobClas           jobType
0         Others                  
1  Project Mgmt.   Project Manager
2   Data Science  Business Analyst
3   Data Science    Data Scientist
4   Data Science      Data Analyst
5   Data Science     Data Engineer
6   Data Science             ML/AI

In [20]:
res_df=pd.merge(res_df,cjob_df,on="jobType", how='left')

In [21]:
res_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 45860 entries, 0 to 45859
Data columns (total 17 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   size          45860 non-null  object 
 1   sector        45860 non-null  object 
 2   industry      45860 non-null  object 
 3   type          45860 non-null  object 
 4   ccode         45860 non-null  object 
 5   cname         45860 non-null  object 
 6   exp           45860 non-null  object 
 7   level         45860 non-null  object 
 8   jobType       45860 non-null  object 
 9   total         45860 non-null  int64  
 10  salLow_min    1518 non-null   float64
 11  salLow_max    1518 non-null   float64
 12  salLow_mean   1518 non-null   float64
 13  salHigh_min   1518 non-null   float64
 14  salHigh_max   1518 non-null   float64
 15  salHigh_mean  1518 non-null   float64
 16  jobClas       45860 non-null  object 
dtypes: float64(6), int64(1), object(10)
memory usage: 6.3+ MB


## 4. Análisis de tendencias en los datos (EDA)
________________

### PIE CHARTS

In [22]:
    res_df.loc[res_df.jobType=="", 'jobType'] = "Others"
    fig = px.pie(res_df, values="total", names='jobType', title="Oferta de empleo por tipo (JobTitle)")
    fig.show()

In [24]:

fig = px.sunburst(res_df, path=['jobClas', 'jobType'], values=(res_df['total']/res_df['total'].sum())*100 )
fig2 = px.sunburst(res_df, path=['jobClas', 'jobType'], values=res_df['total'])
fig.show()
fig2.show()






In [27]:
df3 = res_df[res_df.industry!="N/A"]
fig3 = px.sunburst(df3, path=['jobClas', 'jobType','industry'], values=df3['total'])
fig3.show()

In [ ]:
    res_df.loc[res_df.jobType=="", 'jobType'] = "Others"
    fig = px.pie(res_df, values="total", names='jobType', title="OFERTA DE EMPLEO POR TIPO")
    fig.show()

### GRÁFICOS P

In [94]:
GDjobs_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 155890 entries, 0 to 165284
Data columns (total 24 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   empSize      155890 non-null  object        
 1   easyApply    155890 non-null  bool          
 2   empName      155890 non-null  object        
 3   jobTitle     155890 non-null  object        
 4   jobDate      155890 non-null  datetime64[ns]
 5   salHigh      3645 non-null    float64       
 6   salLow       3645 non-null    float64       
 7   jobDesc      155890 non-null  object        
 8   jobSource    155890 non-null  object        
 9   country      155890 non-null  object        
 10  lat          155890 non-null  float64       
 11  long         155890 non-null  float64       
 12  location     155890 non-null  object        
 13  foundedYear  155890 non-null  int64         
 14  industry     155890 non-null  object        
 15  revenue      155890 non-null  obje

In [92]:
GDjobs_df.jobDate.max()

Timestamp('2019-11-10 00:00:00')

###   ANÁLISIS POR 

In [84]:
    fig = px.area(res_df[res_df.level!='N/A'],x="size",y="total",color="jobType",line_group="jobType")
    fig.show()

## Como se distribuye la oferta en el mundo?

In [76]:
#Dataframe resumen con la información mundial 
rworld_df = (res_df.groupby(['cname','jobType', 'jobClas'])['total'].sum()).to_frame()

In [77]:


rworld_df.reset_index(inplace=True)

In [78]:
rworld_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 657 entries, 0 to 656
Data columns (total 4 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   cname    657 non-null    object
 1   jobType  657 non-null    object
 2   jobClas  657 non-null    object
 3   total    657 non-null    int64 
dtypes: int64(1), object(3)
memory usage: 20.7+ KB


In [72]:
#df = px.data.gapminder()

In [82]:
dsub = rworld_df[rworld_df.jobClas=='Data Science']
fig = px.scatter_geo(rworld_df, locations="cname", locationmode="country names" ,color="jobType", hover_name="cname", size="total", projection="natural earth"
 #,scope="europe"
 )
fig.show()

## OTROS PENDIENTES

Análisis de Tools  
Análisis por nombre de empresa
Histogramas por salario medio y por antiguedad y box plot

In [ ]:

#Revisión de la columna Job Description para ubicar en ella las principales tecnologias requeridas
# BUSINESS INTELLIGENCE TOOLS
# Tableau, MicroStrategy, PowerBI(Microsoft), QLIKSense, Business Objects(SAP), COGNOS(IBM), EXCEL
# DATA ANALYSIS TOOLS (LANGUAGES)
# Python, R, SAS, MATLAB, SQL
# MACHINE LEARNING
# ML (Machine Learning), Time Series, Predic, NLP, SciKit, PyTorch, AML
# GESTION PROYECTOS
# Agile, Scrum, PMO, Microsoft Project, JIRA, TRELLO
# CLOUD
# AWS, AZURE, Google Cloud  (y CLOUD en general)
# BIG DB TOOLS (Open Source)
# SPARK, CASSANDRA, HADOOP, HIVE, MONGO DB
# OTROS 
# PANDAS, API, JSON, CRM
# Aumentar el despliegue del texto en pantalla de las columnas (especialmente para el caso de JobDesc)
#pd.set_option('display.max_colwidth', 2000)


In [ ]:
#GDjobs_df[(GDjobs_df["jobDesc"].str.contains("Cloud ML",na=False))].count()
#GDjobs_df[(GDjobs_df["jobTitle"].str.contains("Project Manager",na=False))& (GDjobs_df["jobType"]== "")]

In [ ]:
#GDjobs_df["empName"].nunique()      # Hay 30567 empresas en total. Pendiente analizar top empresas y sector en cada perfil